File: Core kernel razonamiento repo para Yaiwes/LangChain-Task-Steering/packages/typescript/src/adapter.ts
/**
 * Adapter to use an AgentMiddleware-shaped object as a TaskMiddleware.
 */

import {
  TaskMiddleware,
  type ModelRequest,
  type ModelResponse,
  type ModelCallHandler,
  type ToolCallRequest,
  type ToolCallHandler,
  type ToolMessageResult,
  type CommandResult,
  type ToolLike,
} from './types.js'

/**
 * Shape of an agent-level middleware that the adapter can wrap.
 * Only the hooks that overlap with TaskMiddleware are required.
 */
export interface AgentMiddlewareLike {
  wrapModelCall?(request: ModelRequest, handler: ModelCallHandler): ModelResponse
  wrapToolCall?(
    request: ToolCallRequest,
    handler: ToolCallHandler
  ): ToolMessageResult | CommandResult
  tools?: ToolLike[]
}

/**
 * Wraps a standard agent-level middleware so it can be used at task scope.
 *
 * Forwards the hooks that overlap between the two interfaces:
 *
 * - `wrapModelCall` → delegates to the inner middleware
 * - `wrapToolCall`  → delegates to the inner middleware
 * - `tools`         → exposes the inner middleware's tools (merged
 *   with the task's own tools by `TaskSteeringMiddleware`)
 *
 * Agent-level hooks (`beforeAgent`, `afterAgent`) have no task-scope
 * equivalent and are **not** forwarded — use `onStart` / `onComplete`
 * for task lifecycle events.
 *
 * @example
 * ```ts
 * import { AgentMiddlewareAdapter } from 'langchain-task-steering'
 *
 * const task: Task = {
 *   name: 'research',
 *   instruction: '...',
 *   tools: [searchTool],
 *   middleware: new AgentMiddlewareAdapter(new SummarizationMiddleware()),
 * }
 * ```
 */
export class AgentMiddlewareAdapter extends TaskMiddleware {
  private readonly _inner: AgentMiddlewareLike

  /** Tools contributed by the inner middleware. */
  readonly tools: ToolLike[]

  constructor(inner: AgentMiddlewareLike) {
    super()
    this._inner = inner
    this.tools = inner.tools ? [...inner.tools] : []

    // Wire up optional hooks only if the inner middleware implements them
    if (inner.wrapModelCall) {
      this.wrapModelCall = (request: ModelRequest, handler: ModelCallHandler): ModelResponse => {
        return inner.wrapModelCall!(request, handler)
      }
    }

    if (inner.wrapToolCall) {
      this.wrapToolCall = (
        request: ToolCallRequest,
        handler: ToolCallHandler
      ): ToolMessageResult | CommandResult => {
        return inner.wrapToolCall!(request, handler)
      }
    }
  }
}

File: ➡️📂Core kernel Yaiwes/LangChain-Task-Steering/packages/python/src/langchain_task_steering/adapter.py
"""Adapter to use an AgentMiddleware as a TaskMiddleware."""

from typing import Any

from langchain.agents.middleware import AgentMiddleware

from ._hooks import WRAP_HOOK_PAIRS, overrides_base
from .types import TaskMiddleware


class AgentMiddlewareAdapter(TaskMiddleware):
    """Wraps a standard ``AgentMiddleware`` so it can be used at task scope.

    Wrap-style hooks (``wrap_*`` / ``awrap_*``) are discovered dynamically
    from ``AgentMiddleware`` at import time.  Only hooks the inner middleware
    actually overrides are forwarded; the rest fall through to the handler.

    This means new hooks added to ``AgentMiddleware`` in future LangChain
    versions are picked up automatically — no code changes required here.

    Additionally forwards:

    - ``tools``        → inner middleware's tools (scoped by ``TaskSteeringMiddleware``)
    - ``state_schema`` → inner middleware's state schema (merged into agent state)

    Agent-level hooks (``before_agent``, ``after_agent``) are **not** forwarded
    — use ``on_start`` / ``on_complete`` for task lifecycle events.

    Example::

        from langchain.agents.middleware import SummarizationMiddleware
        from langchain_task_steering import Task, AgentMiddlewareAdapter

        Task(
            name="research",
            instruction="...",
            tools=[search_tool],
            middleware=AgentMiddlewareAdapter(SummarizationMiddleware()),
        )
    """

    def __init__(self, inner: AgentMiddleware) -> None:
        super().__init__()
        self._inner = inner

        inner_tools = getattr(inner, "tools", None)
        self.tools: list = list(inner_tools) if inner_tools else []

        inner_schema = getattr(inner, "state_schema", None)
        if inner_schema is not None:
            self.state_schema = inner_schema

        # Dynamically bind forwarding methods for every discovered
        # wrap-style hook the inner middleware actually overrides.
        for sync_name, async_name in WRAP_HOOK_PAIRS:
            self._bind_sync_hook(sync_name)
            if async_name:
                self._bind_async_hook(async_name)

    def _bind_sync_hook(self, method_name: str) -> None:
        inner = self._inner
        has_override = overrides_base(inner, method_name)

        def forwarder(request, handler, _m=method_name, _has=has_override):
            if _has:
                return getattr(inner, _m)(request, handler)
            return handler(request)

        setattr(self, method_name, forwarder)

    def _bind_async_hook(self, method_name: str) -> None:
        inner = self._inner
        has_override = overrides_base(inner, method_name)

        async def forwarder(request, handler, _m=method_name, _has=has_override):
            if _has:
                return await getattr(inner, _m)(request, handler)
            return await handler(request)

        setattr(self, method_name, forwarder)
